### ISSUE TITLE
[Bug]: "bwrap: Can't find source path …/doc/by-app/…: Permission denied" on Fedora when using BTRFS Assistant

### ISSUE THREAD

**Issue by AdivonSlav** [flatpak/flatpak#5860]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.8

### What Linux distribution are you using?

Fedora Linux

### Linux distribution version

40 KDE Edition

### What architecture are you using?

x86_64

### How to reproduce

Attempt to launch a flatpak app e.g. with `flatpak run`

### Expected Behavior

The app being run to actually open

### Actual Behavior

No matter the app, I will get a similar error to this
`bwrap: Can't find source path /run/user/1000/doc/by-app/org.qbittorrent.qBittorrent: Permission denied`

### Additional Information

Found a few opened issues which were unfortunately closed with no apparent solution on the Flatpak GitHub. One thing that I know fixes this (at least in my case) temporarily is running `sudo killall xdg-document-portal` After that, all the apps launch just fine.

I'm still not able to pinpoint exactly what causes this to occur.

Some additional system info:
- Linux kernel 6.9.6
- KDE Plasma 6.1.1 Wayland

**Comment by nothingto** [comment]
I can add a tidbit that it seems that everyone who is having this issue (me included, apps do not run when pressed or take 5+ minutes but when run through terminal get this same error for other apps), this is happening to people using Fedora. I'm using fedora 40 GNOME. I initially thought this was a GNOME issue, since I didn't ahve this issue with KDE. However, I almost exclusively started relying on flatpaks so I just guess I noticed it now.

I am not particularly sure about the Fedora part. It might be other distros. I might install Arch again and start testing there.

**Comment by AdivonSlav** [comment]
Any update on this, any triage, anything? Happened today again

**Comment by AdivonSlav** [comment]
One thing I managed to figure out is that this happens after I open BTRFS Assistant weirdly enough, which aligns with what other people report online about this being linked to BRFS snapshots somehow.

**Comment by AdivonSlav** [comment]
https://discussion.fedoraproject.org/t/btrfs-assistant-increasing-flatpak-apps-boot-time/120497/8

According to this at least, seems to be an issue only on Fedora and its spins.

**Comment by smcv** [comment]
What version of `xdg-desktop-portal` is installed on the affected systems?

The directories `$XDG_RUNTIME_DIR/doc/by-app/…` are provided by the `xdg-documents-portal`, which mounts a FUSE filesystem on `$XDG_RUNTIME_DIR/doc`. This is required functionality for the file chooser portal. The file chooser is the component that is responsible for File → Open and File → Save As… dialogs running un-sandboxed on the host system, and if the user selects a file to open, transparently making that file available to the appropriate app via its app-specific "view" below `$XDG_RUNTIME_DIR/doc/by-app/…`.

If `$XDG_RUNTIME_DIR/doc/by-app/…` is broken, then it's normal that Flatpak apps will fail to start. It should be technically possible to start them with `flatpak run --no-documents-portal …`, but they will have *severely* reduced functionality (many apps will be effectively unusable, because opening and saving files won't be possible).

You should also be able to access `$XDG_RUNTIME_DIR/doc/by-app/…` directly, from outside the Flatpak sandbox, by using a shell or a file manager. On affected systems, if you do that, do you get a similar `Permission denied` error?

Some possible root causes:

* This could be a bug in `xdg-documents-portal` (which is part of https://github.com/flatpak/xdg-desktop-portal, not Flatpak) when running on a btrfs filesystem where snapshots exist.
* Or if this is Fedora-specific, that suggests it could be a SELinux policy denying our access to something that we need to access. Are SELinux denials logged in the Journal or the audit log when this happens?

If someone can put together steps-to-reproduce starting with a Fedora live image, or a fresh Fedora installation into a virtual machine, then that would probably be helpful.

**Comment by Heavygrass** [comment]
@smcv smcv
> What version of `xdg-desktop-portal` is installed on the affected systems?

How to determine version of `xdg-desktop-portal`?

I'm observing what seems to be a relevant errors for all Flatpak apps on Ubuntu 24.04 with a bunch of BTRFS subvolumes/snapshots:
`/usr/bin/flatpak run --branch=stable --arch=x86_64 --command=torbrowser-launcher --file-forwarding org.torproject.torbrowser-launcher @@u %u @@
bwrap: Can't find source path /run/user/1000/doc/by-app/org.torproject.torbrowser-launcher: No such file or directory.`

Folder /run/user/1000/doc is EMPTY when such errors start.
Restarting xdg-documents-portal resolves the issue.
Sorry if I'm reporting to a wrong thread.

**Comment by smcv** [comment]
> bwrap: Can't find source path /run/user/1000/doc/by-app/org.torproject.torbrowser-launcher: No such file or directory.`

This is a different error (this issue says "Permission denied", yours is "No such file or directory") and you say you are not using BTRFS Assistant, so please report it separately.

xdg-documents-portal is part of https://github.com/flatpak/xdg-desktop-portal/ so please report it there, rather than as a Flatpak issue (Flatpak depends on it but it is not part of Flatpak).

Also please check the system log (systemd Journal) for any warning or error messages from xdg-documents-portal at around the time that you saw this error.

> How to determine version of `xdg-desktop-portal`?

It depends on your OS distribution. On Ubuntu: `dpkg-query -W xdg-desktop-portal`

**Comment by hmnd** [comment]
As a recent Fedora convert, I'm experiencing this issue now too, just after launching `btrfs-assistant` for the first time. I am running version 1.18.4 of `xdg-desktop-portal`

**Comment by da-anda** [comment]
can confirm the issue, and that the flatpaks start again when lauchned with `--no-documents-portal`. I as well am on Fedora and hat `btrfs-assistant` running. But flatpaks also refuse to start for me when a fstab mounted network share is not available. Will double check if it is the same issue (documents portal) once it happens again

**Comment by Batwam** [comment]
I just faced the same issue and can confirm that the app started again after running `sudo killall xdg-document-portal`. I am also running btrfs-assistant.

**Comment by ravilock** [comment]
There are mentions of this same problem on `xdg-desktop-portal` [here](https://github.com/flatpak/xdg-desktop-portal/issues/1410), and on `btrfs-assitant` [here](https://gitlab.com/btrfs-assistant/btrfs-assistant/-/issues/105).

cc @Batwam

**Comment by chrisawi** [comment]
This isn't a Flatpak bug. btrfs-assistant is causing a second set of xdg-desktop-portal processes to be spawned as root (only on Fedora, for whatever reason).

**Comment by SAJewers** [comment]
> This isn't a Flatpak bug. btrfs-assistant is causing a second set of xdg-desktop-portal processes to be spawned as root (only on Fedora, for whatever reason).

In that case, has a bug been filed on fedora's side for that?

**Comment by Batwam** [comment]
In my case, this was happening in Ubuntu as well

**Comment by lihaohong6** [comment]
> > This isn't a Flatpak bug. btrfs-assistant is causing a second set of xdg-desktop-portal processes to be spawned as root (only on Fedora, for whatever reason).
> 
> In that case, has a bug been filed on fedora's side for that?

Yes. Here's the [bug report](https://bugzilla.redhat.com/show_bug.cgi?id=2365186).

**Comment by pedropaulosuzuki** [comment]
Same problem here, but with Flatpak Builder. Using Ubuntu.

```
Building module gnuchess in /home/ubuntu/Documents/GNOME Chess/.flatpak-builder/build/gnuchess-4
========================================================================
bwrap: Can't find source path /home/ubuntu/Documents/GNOME Chess/build/files: Permission denied
Error: module gnuchess: Child process exited with code 1
```

**Comment by SAJewers** [comment]
Is Ubuntu using the latest version? the btrfs-assistant devs claim to have fixed it: https://gitlab.com/btrfs-assistant/btrfs-assistant/-/issues/105

**Comment by lihaohong6** [comment]
@pedropaulosuzuki This has been resolved on Fedora after upgrading to btrfs-assistant 2.2. Ubuntu is [still stuck on 2.1](https://packages.ubuntu.com/search?suite=all&searchon=names&keywords=btrfs-assistant). You need to contact Ubuntu packagers and ask them to update btrfs-assistant. Alternatively, you can update btrfs-assistant yourself, though it'll likely cause issues in your system.

**Comment by OfficialMuffin** [comment]
In my case I simply had to make sure bwrap permissions were set properly with:
```
sudo chown root:root /usr/bin/bwrap
sudo chmod 0755 /usr/bin/bwrap
```
Source: https://askubuntu.com/questions/1489166/can-t-run-any-flatpak-app
